sway-lib-std/src/u128.sw

```diff
@@ -0,0 +1,310 @@
+library u128;
+
+use core::num::*;
+use ::assert::assert;
+use ::result::Result;
+
+/// The 128-bit unsigned integer type.
+/// Represented as two 64-bit components: `(upper, lower)`, where `value = (upper << 64) + lower`.
+pub struct U128 {
+    upper: u64,
+    lower: u64,
+}
+
+pub trait From {
+    /// Function for creating U128 from its u64 components.
+    pub fn from(h: u64, l: u64) -> Self;
+} {
+    fn into(v: U128) -> (u64, u64) {
+        (v.upper, v.lower)
+    }
+}
+
+impl From for U128 {
+    pub fn from(h: u64, l: u64) -> U128 {
+        U128 {
+            upper: h,
+            lower: l,
+        }
+    }
+}
+
+impl core::ops::Eq for U128 {
+    pub fn eq(self, other: Self) -> bool {
+        self.lower == other.lower && self.upper == other.upper
+    }
+}
+
+impl core::ops::Ord for U128 {
+    pub fn gt(self, other: Self) -> bool {
+        self.upper > other.upper || self.upper == other.upper && self.lower > other.lower
+    }
+
+    pub fn lt(self, other: Self) -> bool {
+        self.upper < other.upper || self.upper == other.upper && self.lower < other.lower
+    }
+}
+
+// TODO this doesn't work?
+// impl core::ops::OrdEq for U128 {
+// }
```

Comment:
If I uncomment these lines I get the following error:
```console
error
  --> sway-lib-std/src/u128.sw:49:1
   |
47 |   
48 |   // TODO this doesn't work?
49 | / impl core::ops::OrdEq for U128 {
50 | | }
   | |_^ The trait "Ord" is not implemented for type "struct U128<u64, u64>"
51 |   
52 |   fn disable_overflow() {
   |
```

even though `Ord` is clearly defined above. I don't think if this is because it's specifically expecting trait `Ord` instead of `core::ops::Ord`, or because or the order of definitions.

Reply:
cc @otrho @emilyaherbert

Reply:
#1667